extensions/ext/GLSL_EXT_integer_dot_product.txt

```diff
@@ -0,0 +1,288 @@
+      int dotAccSatEXT(i16vec3 a, u16vec3 b, int c)
+      int dotAccSatEXT(u16vec3 a, i16vec3 b, int c)
+      int dotAccSatEXT(i16vec3 a, i16vec3 b, int c)
+
+      uint dotAccSatEXT(u16vec4 a, u16vec4 b, uint c)
+      int dotAccSatEXT(i16vec4 a, u16vec4 b, int c)
+      int dotAccSatEXT(u16vec4 a, i16vec4 b, int c)
+      int dotAccSatEXT(i16vec4 a, i16vec4 b, int c)
+
+      If GL_EXT_shader_explicit_arithmetic_types_int64 is enabled:
+
+      uint64_t dotAccSatEXT(u64vec2 a, u64vec2 b, uint64_t c)
+      int64_t dotAccSatEXT(i64vec2 a, u64vec2 b, int64_t c)
+      int64_t dotAccSatEXT(u64vec2 a, i64vec2 b, int64_t c)
+      int64_t dotAccSatEXT(i64vec2 a, i64vec2 b, int64_t c)
+
+      uint64_t dotAccSatEXT(u64vec3 a, u64vec3 b, uint64_t c)
+      int64_t dotAccSatEXT(i64vec3 a, u64vec3 b, int64_t c)
+      int64_t dotAccSatEXT(u64vec3 a, i64vec3 b, int64_t c)
+      int64_t dotAccSatEXT(i64vec3 a, i64vec3 b, int64_t c)
+
+      uint64_t dotAccSatEXT(u64vec4 a, u64vec4 b, uint64_t c)
+      int64_t dotAccSatEXT(i64vec4 a, u64vec4 b, int64_t c)
+      int64_t dotAccSatEXT(u64vec4 a, i64vec4 b, int64_t c)
+      int64_t dotAccSatEXT(i64vec4 a, i64vec4 b, int64_t c)
+
+      The function dotAccSatEXT() computes the dot product of the two input
+      vectors <a> and <b> and then returns the result of the saturating addtion
+      with the accumulator <c>.
+
+      All components of the input vectors are extended to the bit width of
+      the result's type if the bit width of the result's type is larger than
+      that of the components of the input vectors, using sign-extension for
+      signed inputs and zero-extension for unsigned inputs. The extended input
+      vectors are then multiplied component-wise and all components of the
+      vector resulting from the component-wise multiplication are added
+      together. Finally, the resulting sum is added to the input accumulator.
+      This final addition is saturating.
+
+      If any of the multiplications or additions, with the exception of the
+      final accumulation, overflow or underflow, the result of the instruction
+      is undefined.
+
+      Where vectors are packed into 32-bit integers, vector components follow
+      byte significance order with the lowest-numbered component stored in the
+      least significant byte.
+Issues
+
+    (1) What overloads should this extension support?
+
+    RESOLVED: The VK_KHR_shader_integer_dot_product extension requires that all
+    data types accepted by the Vulkan implementation be usable with dot product
+    instructions. This extension provides overloads for all possible input
+    vector types. The return values are 32-bit integers, except for dot
+    products operating on vectors of 64-bit integers, as this is expected to be
+    the common case and it is easy for compilers to fold any subsequent
+    truncations or zero/sign-extensions into the dot product instruction as an
+    optimization.
```

Comment:
I see that this partially addresses my comment above, but it doesn't completely address it because the precision of the intermediate results is required to match the output precision, so a 32-bit operation and a truncate is not the same as a 16-bit operation.

Reply:
I think it is because we require "wrap-around behaviour". Following the SPIR-V wording, the 32-bit operation would provide the correct result R.

Reply:
Yes, I think my original comment is just wrong, the operation + truncate does provide the same as the smaller operation, so this does resolve my comment above.